src/compiler/checker.ts

```diff
@@ -6814,6 +6904,22 @@ module ts {
             return undefinedType;
         }
 
+        function checkAwaitExpression(node: AwaitExpression): Type {
+            // Grammar checking
+            if (!(node.parserContextFlags & ParserContextFlags.Await)) {
+                var parameter = getContainingParameter(node);
+                if (parameter && parameter.parserContextFlags & ParserContextFlags.Await) {
+                    grammarErrorAfterFirstToken(node, Diagnostics._0_expression_is_not_allowed_in_an_initializer, "await");

```

Comment:
If we do not allow await expressions on initializers, then I see no reason to have an AsyncParameter context. Maybe we should make things simpler by just treating all await tokens in an async context as a keyword.